using Boto3 in s3.py

s-hertel opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/s3.py

##### ANSIBLE VERSION
```
ansible 2.3.0 (boto3-s3 13631c256d) last updated 2017/02/16 13:36:21 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
Updating S3 since all new AWS module pull requests are expected to use boto3.  This will also fix signature version bugs (eg #21200)

Closes #23757

mattclay commented:
CI failure due to PEP 8 issue:

```
2017-02-24 22:01:00 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/s3.py: Passes current rule set. Remove from legacy list (test/sanity/pep8/legacy-files.txt).
```

This means you've resolved outstanding PEP 8 issues in your PR and the file no longer needs to be listed in legacy files list. Just remove it from that file as part of your PR.

You can run PEP 8 tests locally with `make pep8`.

mattclay commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -706,41 +774,46 @@ def main():
+                        endpoint=walrus
+                        **aws_connect_kwargs
+        )
+                        
```
CI failure due to PEP 8 issue:

```
2017-02-27 20:21:42 ERROR: PEP 8: lib/ansible/modules/cloud/amazon/s3.py:807:1: W293 blank line contains whitespace (current)
```

The PEP 8 tests can be run locally with `make pep8`.

mattclay commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -261,139 +273,161 @@
+from ansible.module_utils.ec2 import ec2_argument_spec, get_aws_connection_info, boto3_conn, HAS_BOTO3
 
 try:
-    import boto
```
Since this module no longer depends on boto, you should update the [unit tests](https://github.com/ansible/ansible/blob/devel/test/units/modules/cloud/amazon/test_s3.py#L4) to reflect that.

You'll also want to remove `boto` from the [unit test requirements](https://github.com/ansible/ansible/blob/devel/test/runner/requirements/units.txt#L1).

s-hertel replied:
@mattclay Fixed!

mattclay commented:
My review was just for the unit test changes.

mattclay commented:
CI failure in unit tests on python 2.6:

```
2017-03-01 02:10:06 >       self.assertIn("bla.blubb", str(actual._endpoint))
2017-03-01 02:10:06 E       AttributeError: 'TestUrlparse' object has no attribute 'assertIn'
2017-03-01 02:10:06 
2017-03-01 02:10:06 test/units/modules/cloud/amazon/test_s3.py:43: AttributeError
```

mattclay commented:
CI failure in unit test `test_get_s3_connection`:

```
>       actual = s3.get_s3_connection(None, aws_connect_kwargs, location, rgw, s3_url)
E       TypeError: get_s3_connection() takes exactly 6 arguments (5 given)

test/units/modules/cloud/amazon/test_s3.py:33: TypeError
```

s-hertel committed:
replace boto with boto3 for the s3 module

make s3 pep8 and remove from legacy files

fix s3 unit tests

s-hertel committed:
fix indentation

michael-dev2rights commented:
Summary:

Given an 11M upload file the attached playbook fails with the message that:
"Files uploaded with multipart of s3 are not supported with checksum, unable to compute checksum."

What I'd expect to happen:

Ansible should normally succeed in doing the same action twice.  

What actually happens:

Uploading the same file twice fails.

mikedlr committed:
s3 module - if we can't create an MD5 sum return None and always upload file

ryansb commented:
Would also close out https://github.com/ansible/ansible-modules-core/issues/3140

roadmapper commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -129,10 +132,9 @@
+     - "AWS region to create the bucket in. If not set then the value of the AWS_REGION and EC2_REGION environment variables
+       are checked, followed by the aws_region and ec2_region settings in the Boto config file. If none of those are set the
+       region defaults to the S3 Location: US Standard. Prior to ansible 1.8 this parameter could be specified but had no effect."
     required: false
```
Could the region be a required parameter? It seems like Location.DEFAULT is really vestigial after switching to boto3. Looks like there was an old issue in ansible-modules-core (https://github.com/ansible/ansible-modules-core/issues/964#issuecomment-86552229) where the Location.DEFAULT causing a NoneType was discussed. If it must stay as not required, would it make sense to hardcode this to us-east-1?

s-hertel replied:
I'm thrilled to have another set of eyes on this. I think you're right. When I was working on this I couldn't actually find Location.Default in Boto3.  And since this module needs to be backwards compatibility I think I'll just keep the default hard coded to us-east-1 like is happening now
https://github.com/s-hertel/ansible/blob/8f44e7482c2aa9e31139e1a915d016587009aefa/lib/ansible/modules/cloud/amazon/s3.py#L584-L592. When I come back to this to tidy it up again I'll be sure to fix up the docs and remove any other traces of Default.Location/comments regarding it. :) Good catch!

Edit:
Oh, but I don't think we can require region since it wasn't required before and we should be able to use environment variables instead of needing to specify region in the playbook.

s-hertel committed:
remove Location.DEFAULT which isn't used in boto3 and tidy up the docs

s-hertel committed:
pep8

mattclay commented:
CI failure in multiple unit tests: https://app.shippable.com/github/ansible/ansible/runs/26413/2/tests

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/28159/2/tests

s-hertel committed:
s3: remove default: null, empty aliases, and required: false from documentation

fix incorrectly documented defaults

s-hertel committed:
Porting s3 to boto3. Simplify some logic and remove unused imports

s-hertel committed:
Fix s3 module variables

s-hertel committed:
Fix a typo in s3 module and remove from pep8 legacy files

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,140 @@
-    key_check = bucket.get_key(obj, version_id=version)
+def keysum(module, s3, bucket, obj, version=None):
+    if version:
+        key_check = s3.get_object(Bucket=bucket, Key=obj, VersionId=version)
```
`get_object` will get the full item (including the body) - instead I think we should use `head_object` here as it'll still give back the etag without getting the full item. http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Client.head_object

To test this, you could make a 1-2GB object and then try this codepath. It should take a long time, and consume 1-2GB of memory.

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,140 @@
-    etag_multipart = '-' in md5_remote # Check for multipart, etag is not md5
+    md5_remote = key_check['ETag'][1:-1]
+    etag_multipart = '-' in md5_remote  # Check for multipart, etag is not md5
     if etag_multipart is True:
```
this would be equivalent of `if '-' in md5_remote:` since `etag_multipart` is never used other than this conditional.

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,140 @@
-        location = Location.DEFAULT
+        location = 'us-east-1'
+    configuration = {}
+    if location != ('us-east-1' or None):
```
This conditional is never going to evaluate properly, since `('us-east-1' or None)` evaluates to `True`, so the only way this evaluates as wrong is if location == True. I think what you mean is `location not in ('us-east-1', None):`

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,140 @@
-            module.fail_json(msg=str(e), exception=traceback.format_exc())
-    try:
-        bucket.delete()
+        objects = s3.list_objects(Bucket=bucket)
```
I have a concern on pagination here, `list_objects` is going to default to 100 or 500 items, which is fine most of the time (it was like this prior) but would make a nice enhancement.

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -405,63 +382,82 @@ def path_check(path):
-
-        key.set_contents_from_filename(src, encrypt_key=encrypt, headers=headers)
+            extra = {'Metadata': {}}
+            for meta_key in metadata:
```
An equivalent to this would be `extra['Metadata'].update(metadata)`

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -405,63 +382,82 @@ def path_check(path):
+        if version:
+            contents = to_text(s3.get_object(Bucket=bucket, Key=obj, VersionId=version)["Body"].read())
+        else:
+            contents = to_text(s3.get_object(Bucket=bucket, Key=obj)["Body"].read())
```
I think you need `to_native` here.

s-hertel committed:
s3: add pagination for listing objects.

Fix logic and use head_object instead of get_object for efficiency.

Fix typo in unit test.

ryansb commented on test/runner/requirements/units.txt:
```diff
@@ -1,4 +1,3 @@
-boto
```
Are there really no other tests that need boto at all?

s-hertel replied:
It's because there aren't very many tests. If there are tests written for old boto modules this will need to be added back.
https://github.com/ansible/ansible/search?utf8=%E2%9C%93&q=boto+path%3Atest%2Funits%2F&type=

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,139 @@
-    if location is None:
-        location = Location.DEFAULT
+    configuration = {}
+    if location not in ('us-east-1' or None):
```
Looks like another weird conditional slipped through, this should be `if location not in ('us-east-1', None):`

s-hertel replied:
Sorry, derp.

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -405,63 +381,80 @@ def path_check(path):
-            key.set_acl(acl)
-        url = key.generate_url(expiry)
+            s3.put_object_acl(ACL=acl, Bucket=bucket, Key=obj)
+        url = url = s3.generate_presigned_url(ClientMethod='put_object',
```
Technically this sets the proper value for `url` but it's doing so twice. s/url = url/url/

s-hertel committed:
Fix pagination to maintain backwards compatibility.

Fix incorrect conditional.

Remove redundant variable assignment.

Fix s3 list_object pagination to return all pages

ryansb commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -326,8 +326,8 @@ def create_bucket(module, s3, bucket, location=None):
     paginator = s3.get_paginator('list_objects')
-    all_keys = [page['Contents'] for page in paginator.paginate(Bucket=bucket)][0]
-    keys = [{'Key': data['Key']} for data in all_keys]
+    all_keys = [page for page in paginator.paginate(Bucket=bucket)][0].get('Contents', [])
```
Won't this just get the contents of the first page? I think you need to flatten this out by extending all `contents` to one list.

willthames commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,139 @@
-    keys = [x.key for x in all_keys]
 
+def list_keys(module, s3, bucket, prefix, marker, max_keys):
+    paginator = s3.get_paginator('list_objects')
```
Use list_objects_v2: http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Client.list_objects_v2

s-hertel replied:
Good catch!!

willthames commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,139 @@
 
+def list_keys(module, s3, bucket, prefix, marker, max_keys):
+    paginator = s3.get_paginator('list_objects')
+    all_keys = [page for page in paginator.paginate(Bucket=bucket)][0].get('Contents', [])
```
These two lines can be simplified (untested, but should work):

```
keys = [data['Key'] for page in paginator.paginate(Bucket=bucket) for data in page.get('Contents', [])]
```

As @ryansb appeared to mention in a comment I can't currently see, this seems only to get the first page.

Edit: I can see the comment in the conversation tab, but not in the code tab.

willthames commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,139 @@
-            module.fail_json(msg=str(e), exception=traceback.format_exc())
-    try:
-        bucket.delete()
+        paginator = s3.get_paginator('list_objects')
```
Comments as for `list_keys`

s-hertel commented:
ready_for_review
Thanks for all the feedback! Pagination is now working correctly and the integration tests are passing locally.

s-hertel committed:
Use the revised List Objects API as recommended.

willthames commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,142 @@
-    all_keys = bucket_object.get_all_keys(prefix=prefix, marker=marker, max_keys=max_keys)
+def paginated_list(s3, bucket):
+    pg = s3.get_paginator('list_objects_v2')
+    for page in pg.paginate(Bucket=bucket):
```
You can do `return pg.paginate(Bucket=bucket).build_full_result()`

Not a blocker though

willthames replied:
Do we need an `AWSRetry` wrapper around this method?

s-hertel replied:
I think I'm going to keep the pagination here as-is so I don't have to filter the results both places where I'm calling this function (since all I want are the key names). Good idea about AWSRetry! I'm making a follow-up PR for that since I don't want to overload this one.

willthames commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,142 @@
-    keys = [x.key for x in all_keys]
 
+def list_keys(module, s3, bucket, prefix, marker, max_keys):
+    keys = [key for key in paginated_list(s3, bucket)]
```
Should this have some exception handling? (I suggest here rather than paginated_list as paginated_list might not be able to handle exceptions if it does the retry)

s-hertel replied:
Totally missed that - good catch.

willthames reviewed (approved):
This can be merged without further changes, but it might be worth considering retries at the very least.

s-hertel committed:
Wrap call to paginated_list in a try/except

Also remembered to allow marker/prefix/max_keys to modify what keys are listed

willthames commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,150 @@
-    try:
-        bucket.delete()
+        # if there are contents then we need to delete them before we can delete the bucket
+        keys = [{'Key': key} for key in paginated_list(s3, **{'Bucket': bucket})]
```
`**{'Bucket': bucket}` is equivalent to `Bucket=bucket`. Please use the latter :)

s-hertel replied:
Fixed :)

willthames commented on lib/ansible/modules/cloud/amazon/s3.py:
```diff
@@ -272,131 +240,150 @@
-    keys = [x.key for x in all_keys]
 
-    module.exit_json(msg="LIST operation complete", s3_keys=keys)
+def list_keys(module, s3, bucket, prefix, marker, max_keys):
```
This function seems to be much more complicated than it needs to be. Does anything call this function with non-trivial values for prefix, marker or max_keys? (I'm guessing previously the function called itself to get the next page).

I would argue for using paginator with build_full_result in list_keys_with_backoff and then the calling functions (`delete_keys` etc.) can just use that directly rather than having to manage the page combination themselves.

s-hertel replied:
Hrm... I'm not sure if I understand what I should be changing.

The user can specify whatever they want for marker or prefix or max_keys, right? I'm not sure what a non-trivial value would be. Previously this function did not call itself to get the next page - pagination wasn't supported at all. So it's true this function has become more complicated. If there's a more elegant way though I'd definitely like to understand.

If I use build_full_result() then I'll have to iterate through that anyway and pull out all the keys so it doesn't seem very different than what I'm doing now. I will implement that if you have a strong preference, but I'm not sure what the benefit is.

I'm a little confused about the comment about delete_keys. list_keys() and delete_bucket() are calling a function that does the pagination. Is the issue that I'm only getting all the keys rather than all the contents?

willthames replied:
By non-trivial I just mean values that aren't None or empty strings. I'm not sure how much user control we expect over those settings but I might not have read the parameters carefully enough.

The following untested somewhat pseudocode illustrates the simpler approach:

```
@AWSRetry(**backoff_params)
def list_keys_with_backoff(connection, bucket):
     pg = connection.get_paginator('list_objects_v2')    
     return [obj['Key'] for obj in pg.paginate(Bucket=bucket).build_full_result()['Objects']]

def list_keys(connection, bucket):
    try:
        return list_keys_with_backoff(connection, bucket)
    except botocore.exceptions.ClientError as e:
        etc...
```

s-hertel committed:
Simplify argument
